File: 08. Trees/02. Binary Search Tree/03. Insert node in a BST/main.js
class TreeNode {
    constructor(val) {
        this.val = val;
        this.left = this.right = null;
    }
}

// Recursive Insert
function insertRecursive(root, key) {
    if (!root) return new TreeNode(key);
    if (key < root.val) root.left = insertRecursive(root.left, key);
    else root.right = insertRecursive(root.right, key);
    return root;
}

// Iterative Insert
function insertIterative(root, key) {
    if (!root) return new TreeNode(key);

    let cur = root, parent = null;
    while (cur) {
        parent = cur;
        if (key < cur.val) cur = cur.left;
        else cur = cur.right;
    }

    if (key < parent.val) parent.left = new TreeNode(key);
    else parent.right = new TreeNode(key);

    return root;
}

// Inorder Traversal (Helper function)
function inorder(root) {
    if (!root) return;
    inorder(root.left);
    process.stdout.write(root.val + " ");
    inorder(root.right);
}

// Main function
function main() {
    let root = new TreeNode(8);
    root = insertRecursive(root, 3);
    root = insertRecursive(root, 10);
    root = insertIterative(root, 6);
    root = insertIterative(root, 14);

    process.stdout.write("Inorder Traversal: ");
    inorder(root);
    console.log();
}

main();

File: 08. Trees/02. Binary Search Tree/03. Insert node in a BST/main.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Recursive Insert
def insert_recursive(root, key):
    if not root:
        return TreeNode(key)
    if key < root.val:
        root.left = insert_recursive(root.left, key)
    else:
        root.right = insert_recursive(root.right, key)
    return root

# Iterative Insert
def insert_iterative(root, key):
    if not root:
        return TreeNode(key)

    cur, parent = root, None
    while cur:
        parent = cur
        if key < cur.val:
            cur = cur.left
        else:
            cur = cur.right

    if key < parent.val:
        parent.left = TreeNode(key)
    else:
        parent.right = TreeNode(key)

    return root

# Inorder Traversal (Helper function)
def inorder(root):
    if root:
        inorder(root.left)
        print(root.val, end=" ")
        inorder(root.right)

# Main function
def main():
    root = TreeNode(8)
    root = insert_recursive(root, 3)
    root = insert_recursive(root, 10)
    root = insert_iterative(root, 6)
    root = insert_iterative(root, 14)

    print("Inorder Traversal:", end=" ")
    inorder(root)
    print()

if __name__ == "__main__":
    main()
